### ISSUE TITLE
Installation

### ISSUE THREAD

**Issue by drtrigon** [AbdealiJK/file-metadata#20]
https://phabricator.wikimedia.org/T136985#2354403
1. `$ pip install file-metadata` does not return useful error messages, had to use `--pre` in order to get a clue
2. `$ pip install --pre file-metadata` does return more useful error messages (I understood that it stopped due to missing deps). However it could be even better, e.g. mentioning the command needed to install the package or a link where to download (depending on OS) - but this is actually the non-pip issues we still have to solve, right? ;) What actually stopped me here is the java message:

> osboxes@osboxes:~$ pip install --pre file-metadata
> Downloading/unpacking file-metadata
>   Downloading file-metadata-0.0.1.dev20160603144066.tar.gz
>   Running setup.py (path:/tmp/pip_build_osboxes/file-metadata/setup.py) egg_info for package file-metadata
>     java version "1.7.0_79"
>     OpenJDK Runtime Environment (IcedTea 2.5.6) (7u79-2.5.6-0ubuntu1.14.04.1)
>     OpenJDK 64-Bit Server VM (build 24.79-b02, mixed mode)
>     error in file-metadata setup command: 'install_requires' must be a string or list of strings containing valid project/version requirement specifiers
>     Complete output from command python setup.py egg_info:
>     java version "1.7.0_79"
> 
> OpenJDK Runtime Environment (IcedTea 2.5.6) (7u79-2.5.6-0ubuntu1.14.04.1)
> 
> OpenJDK 64-Bit Server VM (build 24.79-b02, mixed mode)
> 
> error in file-metadata setup command: 'install_requires' must be a string or list of strings containing valid project/version requirement specifiers
> ---
> 
> Cleaning up...
> Command python setup.py egg_info failed with error code 1 in /tmp/pip_build_osboxes/file-metadata
> Storing debug log for failure in /home/osboxes/.pip/pip.log
> osboxes@osboxes:~$

here I am stuck since I don't see what the issue is exactly:
- is java version "1.7.0_79" needed or the wrongly installed one?
- what do I have to install? (how?)

**Comment by AbdealiLoKo** [comment]
Right, so the `pip install file-metadata` doesn't work because we don't have an official release on pypi yet. (We're still on `.devX` releases). To install the devX (nightly) releases you need `--pre`.

The java issue, isn't a java issue. It just happens to be the step before the actual issue that pip runs. The actual error is:

```
error in file-metadata setup command: 'install_requires' must be a string or list of strings containing valid project/version requirement specifiers
```

I think your pip version is an old one, so it doesn't support conditional statements like `pycolorname>=0.1.0 ; python_version > '2.6'`. So, simplest method is to update the pip, but I could fix it to make it compatible with older pip versions too

Will look into it.

**Comment by AbdealiLoKo** [comment]
Note that the pip in `ubuntu` are really old, so to update the pip to support `; python_version > '2.6'` you'd need to use a virtualenv ... Which makes installing opencv complicated :/

Hence I suggest conda overall ^_^

**Comment by AbdealiLoKo** [comment]
This should be fixed in 382e0813dcbe44af6b0db224b9db98541243cf59 now.
I've made a new devX release for you to test it out with.
I tested it locally with pip==1.0 and setuptools==0.6c11 (These are the versions used in ubuntu-precise) and it worked.

**Comment by drtrigon** [comment]
Thanks! Now I get:

> osboxes@osboxes:~$ pip install --pre file-metadata
> Downloading/unpacking file-metadata
>   Downloading file-metadata-0.0.1.dev20160604021332.tar.gz
>   Running setup.py (path:/tmp/pip_build_osboxes/file-metadata/setup.py) egg_info for package file-metadata
>     error in file-metadata setup command: 'tests_require' must be a string or list of strings containing valid project/version requirement specifiers
>     Complete output from command python setup.py egg_info:
>     error in file-metadata setup command: 'tests_require' must be a string or list of strings containing valid project/version requirement specifiers
> 
> ---
> 
> Cleaning up...
> Command python setup.py egg_info failed with error code 1 in /tmp/pip_build_osboxes/file-metadata
> Storing debug log for failure in /home/osboxes/.pip/pip.log
> osboxes@osboxes:~$

**Comment by AbdealiLoKo** [comment]
Hm, unusual that when I tested it locally it did not raise that issue. I've fixed it on the git repo now. (I had forgotten to remove one problematic case).

Note that a lot has changed since the last version you were trying to install. Now:
- opencv is not being used anymore, skimage is preferred, as skimage can be installed from pip, so opencv does not need to be installed manually
- dlib has become a required dependency
- boost-python is a required dependency which needs to be manually installed because dlib is now mandatory and boost-python has no pypi package.
- imagemagick's wand has become a required dependency
- exiftool doesnt need to be installed manually anymore
- ffmpeg is a required dependency and is installed automatically in linux
- perl is a required dependency that needs to be manually installed

I'll make that "review" version we spoke of on IRC

**Comment by drtrigon** [comment]
Am 6. Juni 2016 08:24:50 MESZ, schrieb AbdealiJK notifications@github.com:

> Hm, unusual that when I tested it locally it did not raise that issue.
> I've fixed it on the git repo now. (I had forgotten to remove one
> problematic case).
> 
> Note that a lot has changed since the last version you were trying to
> install. Now:
> - opencv is not being used anymore, skimage is preferred, as skimage
>   can be installed from pip, so opencv does not need to be installed
>   manually

So you plan to replace opencv completely with skimage?

> - dlib has become a required dependency
> - boost-python is a required dependency which needs to be manually
>   installed because dlib is now mandatory and boost-python has no pypi
>   package.
> - imagemagick's wand has become a required dependency
> - exiftool doesnt need to be installed manually anymore
> - ffmpeg is a required dependency and is installed automatically in
>   linux
> - perl is a required dependency that needs to be manually installed

Why perl? Cause of exiftool?

> I'll make that "review" version we spoke of on IRC

Cool! Thanks! :)

> ---
> 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/AbdealiJK/file-metadata/issues/20#issuecomment-223877589

Dr. Trigon

**Comment by jayvdb** [comment]
It would be nice if the perl dependency can be eliminated, as that makes it more difficult to set up on Windows especially.

**Comment by AbdealiLoKo** [comment]
Issues faced:
-  When doing `pip install --pre file-metadata` and compiling scikit-image, pip said `cython >= 0.21` not found in pypi. probably because it was clashing with the cython in apt-get ?
  - **Solution**: do `pip install --ignore-installed cython` to override the pre-installed cython in apt-get
- In compiling cython, Python.h not found.
  - **Solution**: Do `sudo apt-get install python-dev`. A nice hint about doing this would be awesome.
- Insufficient permissions to write
  - **Solution**: Remind users that sudo is needed if not installing in a virtualenv
- Installing matplotlib failed saying it could not compile "png, freetype"
  - **Solution**: Install them with apt-get and also install pkg-config
- dlib not compiling because cmake not found
  - It says "Cannot find cmake in the path. Please specify its path with --cmake parameter", instructions to install would also be nice.
- numpy "No lapack/blas resource found"
  - Install liblapack-dev libblas-dev
- numpy compiling but no fortran compiler found
  - Install gfortran
- backports.lzma complains "relocation R_X86_64_32S against `_Py_NotImplementedStruct' can not be used when making a shared object; recompile with -fPIC"
  - Python 2.7.2 not supported.
- "ValueError: need more than 0 values to unpack" in file-metadata
  - Not a breaking bug
- `from six.moves.urllib.request import urlopen` not found
  - Need min version of six to be 1.8.0

**Comment by drtrigon** [comment]
Thanks to the IRC session today, solved for me.
See https://phabricator.wikimedia.org/T136985#2358607

**Comment by AbdealiLoKo** [comment]
Made various issues to handle the installation issues.
- Fixed the issue with `six` in 97b5c67bb2d4c8e30e56071d0d939bfe58dc772e
- Created https://github.com/peterjc/backports.lzma/issues/17 for lzma python 2.7.2 compatibility

Will track the other problems in their respective issues

**Comment by drtrigon** [comment]
I think we should put a list of packages that have to be installed prior to the pip page. The list for kubuntu 14.04 is: 

python-dev pkg-config libfreetype6-dev libpng12-dev liblapack-dev libblas-dev gfortran cmake libboost-python-dev liblzma-dev
